lib/src/hooks/useDynamicForm.tsx

```diff
@@ -91,73 +94,83 @@ type OnChangeFormItem = (item: DynamicFormItem) => OnChangeInput;
 type OnChangeInput = (e: InputChange) => void;
 type ShouldRenderFormItem = (sectionKey: string, itemKey: string) => boolean;
 
-export function useDynamicForm(
-    sections: DynamicFormSection[]
-): [() => JSX.Element, any, React.Dispatch<React.SetStateAction<DynamicFormInternalState>>, React.Dispatch<React.SetStateAction<DynamicFormErrors>>, DynamicFormMetaData, boolean] {
-    const initialState: DynamicFormInternalState = {};
-    sections?.map((section) => {
-        initialState[section?.key] = {};
-        section.items?.map((item) => {
-            const { key, value, multi, controlType }: DynamicFormItem = item;
-            let initialValue: any;
+export type FormRenderFunction = () => JSX.Element;
+export type SetDynamicFormState = React.Dispatch<React.SetStateAction<DynamicFormInternalState>>;
+export type SetDynamicFormErrors = React.Dispatch<React.SetStateAction<DynamicFormErrors>>;
+export type SetDynamicFormWarnings = React.Dispatch<React.SetStateAction<DynamicFormErrors>>;
+type isDynamicFormDirty = boolean;
+export type UseDynamicForm = [FormRenderFunction, DynamicFormInternalState, SetDynamicFormState, SetDynamicFormErrors, SetDynamicFormWarnings, DynamicFormMetaData, isDynamicFormDirty];
+export function useDynamicForm(sections: DynamicFormSection[]): UseDynamicForm {
+    const initialState: DynamicFormInternalState = useMemo(() => {
+        let initialFormState: DynamicFormInternalState = {};
+        sections?.map((section) => {
+            initialFormState[section?.key] = {};
+            section.items?.map((item) => {
+                const { key, value, multi, controlType }: DynamicFormItem = item;
+                let initialValue: any;
 
-            switch (controlType) {
-                case "Dropdown": {
-                    if (multi) {
-                        if (value !== null && value !== undefined && Array.isArray(value) && (value as any[]).every((x) => typeof x === "string")) {
-                            initialValue = value as string[];
+                switch (controlType) {
+                    case "Dropdown": {
+                        if (multi) {
+                            if (value !== null && value !== undefined && Array.isArray(value) && (value as any[]).every((x) => typeof x === "string")) {
```

Comment:
if value is an empty array this line will return true, if this is the correct use case then i suggest to use:
```suggestion
                            if (!!(value ?? (Array.isArray(value) && (value as any[]).every((x) => typeof x === "string")))) {
```

Reply:
on top of Omar's comment, could we let the `.isArray` to do the checking for `null` and `undefined`?

```
if (Array.isArray(value) && (value as any[]).every((x) => typeof x === "string")) {
```

Reply:
Omar it's ok to evaluate to true if array is empty as it's a valid state. II will go with Yeoh's comment, I agree the null and undefined check is not needed as will be done is isArray.